File: .github/scripts/detect-changes/src/lib/validation.ts
/**
 * Validation functions for project files.
 *
 * Provides validation capabilities for:
 * - package.json files (detecting real vs placeholder tests)
 * - .nvmrc files (validating semver format)
 * - Test scripts (detecting real test runners)
 * - Dockerfiles (validating structure and rules)
 */

import * as fs from 'fs';
import * as semver from 'semver';
import { validateNoArgInFrom } from './dockerfile-parser.js';
import { ValidationError } from '../utils/errors.js';

/**
 * Real test runner patterns.
 * These indicate actual test frameworks/runners.
 */
const REAL_TEST_RUNNERS = [
  'jest',
  'mocha',
  'tap',
  'pytest',
  'vitest',
  'ava',
  'tape',
  'npm test',
  'node --test',
];

/**
 * Placeholder test patterns.
 * These indicate placeholder/dummy test scripts.
 */
const PLACEHOLDER_PATTERNS = ['echo', 'exit 1', 'exit 0'];

/**
 * Detect if a test script contains real test runners or placeholders.
 *
 * Checks for common test runners (jest, mocha, tap, pytest, vitest, etc.) and
 * identifies placeholder scripts (echo, exit 1, empty strings).
 *
 * @param testScript - The test script command from package.json
 * @returns True if the script appears to be a real test, False if it's a placeholder
 *
 * @example
 * ```typescript
 * hasRealTests('jest')                                    // true
 * hasRealTests('mocha tests/**\/*.test.js')              // true
 * hasRealTests('echo "Error: no test specified"')         // false
 * hasRealTests('exit 1')                                  // false
 * hasRealTests('')                                        // false
 * ```
 */
export function hasRealTests(testScript: string): boolean {
  // Handle null/undefined/empty
  if (!testScript || typeof testScript !== 'string') {
    return false;
  }

  // Trim whitespace
  const normalized = testScript.trim().toLowerCase();

  // Empty after trim
  if (!normalized) {
    return false;
  }

  // Check for placeholder patterns
  for (const pattern of PLACEHOLDER_PATTERNS) {
    if (normalized.includes(pattern)) {
      // If it's just an echo or exit command, it's a placeholder
      // But if it also contains a test runner, it might be a real test
      const hasTestRunner = REAL_TEST_RUNNERS.some((runner) => normalized.includes(runner));
      if (!hasTestRunner) {
        return false;
      }
    }
  }

  // Check for real test runners
  for (const runner of REAL_TEST_RUNNERS) {
    if (normalized.includes(runner)) {
      return true;
    }
  }

  // Default to false if we don't recognize it
  return false;
}

/**
 * Validate a package.json file and detect real vs placeholder tests.
 *
 * Parses package.json and checks if it contains a real test script
 * (using jest, mocha, etc.) or just a placeholder (echo, exit 1).
 *
 * @param packageJsonPath - Absolute path to package.json file
 * @returns True if package.json has real tests, False if placeholder or missing
 * @throws {ValidationError} If file doesn't exist or contains invalid JSON
 *
 * @example
 * ```typescript
 * validatePackageJson('/path/to/package.json')  // true if has real tests
 * ```
 */
export function validatePackageJson(packageJsonPath: string): boolean {
  let content: string;

  try {
    content = fs.readFileSync(packageJsonPath, 'utf8');
  } catch (error) {
    throw new ValidationError(
      `package.json not found at ${packageJsonPath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }

  let packageData: any;
  try {
    // Handle UTF-8 BOM if present
    if (content.charCodeAt(0) === 0xfeff) {
      content = content.slice(1);
    }
    packageData = JSON.parse(content);
  } catch (error) {
    throw new ValidationError(
      `Invalid JSON in ${packageJsonPath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }

  // Check if scripts section exists
  if (!packageData.scripts || typeof packageData.scripts !== 'object') {
    return false;
  }

  // Check if test script exists
  if (!packageData.scripts.test) {
    return false;
  }

  // Check if test script is real
  return hasRealTests(packageData.scripts.test);
}

/**
 * Validate an .nvmrc file for proper semver format.
 *
 * Checks that .nvmrc contains a valid semantic version (with optional 'v' prefix)
 * and rejects LTS aliases like 'lts/*' or 'lts/hydrogen'.
 *
 * @param nvmrcPath - Absolute path to .nvmrc file
 * @returns True if .nvmrc contains valid semver, False if LTS alias or invalid format
 * @throws {ValidationError} If file doesn't exist
 *
 * @example
 * ```typescript
 * validateNvmrc('/path/to/.nvmrc')  // true for "18.20.8"
 * validateNvmrc('/path/to/.nvmrc')  // false for "lts/*"
 * ```
 */
export function validateNvmrc(nvmrcPath: string): boolean {
  let content: string;

  try {
    content = fs.readFileSync(nvmrcPath, 'utf8');
  } catch (error) {
    throw new ValidationError(
      `.nvmrc not found at ${nvmrcPath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }

  // Trim whitespace and normalize line endings
  const version = content.trim();

  // Reject empty content
  if (!version) {
    return false;
  }

  // Reject LTS aliases
  if (version.startsWith('lts/') || version === 'lts/*') {
    return false;
  }

  // Reject named aliases
  const namedAliases = ['node', 'system', 'iojs', 'unstable'];
  if (namedAliases.includes(version.toLowerCase())) {
    return false;
  }

  // Strip 'v' prefix if present
  let versionToCheck = version;
  if (versionToCheck.startsWith('v')) {
    versionToCheck = versionToCheck.slice(1);
  }

  // Check if it's a valid semver or semver range
  // Accept full versions (18.20.8), partial versions (18.20, 18), and ranges
  if (semver.valid(versionToCheck)) {
    return true;
  }

  // Check if it's a valid partial version (coerce will handle "18" or "18.20")
  const coerced = semver.coerce(versionToCheck);
  if (coerced) {
    return true;
  }

  // Invalid format
  return false;
}

/**
 * Validate a Dockerfile for proper structure and rules.
 *
 * Reuses dockerfile-parser validation to check:
 * - No ARG variables in FROM statements
 * - No variable substitution in base images
 *
 * @param dockerfilePath - Absolute path to Dockerfile
 * @returns True if Dockerfile is valid, False otherwise
 * @throws {ValidationError} If file doesn't exist or contains invalid patterns
 *
 * @example
 * ```typescript
 * validateDockerfile('/path/to/Dockerfile')  // true if valid
 * // Throws ValidationError if ARG in FROM
 * ```
 */
export function validateDockerfile(dockerfilePath: string): boolean {
  let content: string;

  try {
    content = fs.readFileSync(dockerfilePath, 'utf8');
  } catch (error) {
    throw new ValidationError(
      `Dockerfile not found at ${dockerfilePath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }

  // Use dockerfile-parser to validate
  try {
    validateNoArgInFrom(content);
    return true;
  } catch (error) {
    if (error instanceof ValidationError) {
      throw error;
    }
    throw new ValidationError(
      `Invalid Dockerfile at ${dockerfilePath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

// Re-export ValidationError for convenience
export { ValidationError };

File: .github/scripts/detect-changes/lib/validation.py
"""
Configuration validation utilities.

Validates:
- package.json structure and test scripts
- .nvmrc format and Node.js version specifications
- Dockerfile syntax and best practices (reusing dockerfile_parser)
"""

import json
import re
from pathlib import Path
from typing import Optional

from dockerfile_parser import ValidationError, validate_no_arg_in_from, parse_from_lines


def validate_package_json(package_json_path: str) -> bool:
    """
    Parse and validate package.json file for real test scripts.

    Args:
        package_json_path: Path to package.json file

    Returns:
        True if package.json has real test scripts, False if placeholder or missing

    Raises:
        ValidationError: If file doesn't exist or contains invalid JSON

    Example:
        >>> validate_package_json("docker/my-service/package.json")
        True
    """
    path = Path(package_json_path)

    if not path.exists():
        raise ValidationError(f"package.json not found: {package_json_path}")

    try:
        with open(path, 'r', encoding='utf-8') as f:
            data = json.load(f)
    except json.JSONDecodeError as e:
        raise ValidationError(f"Invalid JSON in {package_json_path}: {e}")

    # Check for scripts.test
    scripts = data.get('scripts', {})
    test_script = scripts.get('test', '')

    if not test_script:
        return False

    return has_real_tests(test_script)


def validate_nvmrc(nvmrc_path: str) -> bool:
    """
    Validate .nvmrc file format and version specification.

    Args:
        nvmrc_path: Path to .nvmrc file

    Returns:
        True if .nvmrc contains valid semantic version

    Raises:
        ValidationError: If file doesn't exist, is empty, or contains invalid format

    Example:
        >>> validate_nvmrc("docker/my-service/.nvmrc")
        True
    """
    path = Path(nvmrc_path)

    if not path.exists():
        raise ValidationError(f".nvmrc not found: {nvmrc_path}")

    with open(path, 'r', encoding='utf-8') as f:
        content = f.read().strip()

    if not content:
        raise ValidationError(f".nvmrc is empty: {nvmrc_path}")

    # Remove 'v' prefix if present
    version = content.lstrip('v')

    # Check for lts/* format
    if version.lower().startswith('lts/'):
        raise ValidationError(
            f".nvmrc contains lts/* format which is not allowed: {nvmrc_path}. "
            "Use specific semantic version instead (e.g., 18.20.8)"
        )

    # Validate semantic version format
    # Accept X.Y.Z or X.Y
    semver_pattern = r'^\d+\.\d+(\.\d+)?$'
    if not re.match(semver_pattern, version):
        raise ValidationError(
            f".nvmrc contains invalid semver format: {content}. "
            "Expected format: X.Y.Z or X.Y (e.g., 18.20.8)"
        )

    return True


def validate_dockerfile(dockerfile_path: str) -> bool:
    """
    Validate Dockerfile exists and has basic required structure.

    Reuses validation from dockerfile_parser module.

    Args:
        dockerfile_path: Path to Dockerfile

    Returns:
        True if Dockerfile is valid

    Raises:
        ValidationError: If file doesn't exist or fails validation

    Example:
        >>> validate_dockerfile("docker/my-service/Dockerfile")
        True
    """
    path = Path(dockerfile_path)

    if not path.exists():
        raise ValidationError(f"Dockerfile not found: {dockerfile_path}")

    with open(path, 'r', encoding='utf-8') as f:
        content = f.read()

    if not content.strip():
        raise ValidationError(f"Dockerfile is empty: {dockerfile_path}")

    # Check for FROM instruction
    from_lines = parse_from_lines(content)
    if not from_lines:
        raise ValidationError(
            f"Dockerfile missing FROM instruction: {dockerfile_path}"
        )

    # Validate no ARG in FROM
    try:
        validate_no_arg_in_from(content)
    except ValidationError as e:
        # Re-raise with context
        raise ValidationError(f"Dockerfile validation failed for {dockerfile_path}: {e}")

    return True


def has_real_tests(test_script: str) -> bool:
    """
    Detect whether test script is real or a placeholder.

    Args:
        test_script: The test script command from package.json

    Returns:
        True if script appears to run real tests, False for placeholders

    Example:
        >>> has_real_tests("jest")
        True
        >>> has_real_tests("echo \\"Error: no test specified\\" && exit 1")
        False
    """
    if not test_script or not test_script.strip():
        return False

    script = test_script.lower().strip()

    # Placeholder patterns that indicate no real tests
    placeholder_patterns = [
        r'echo.*error.*no.*test',
        r'echo.*no.*test',
        r'^exit\s+1',
        r'echo.*&&.*exit',
    ]

    for pattern in placeholder_patterns:
        if re.search(pattern, script):
            return False

    # Real test runners
    real_test_patterns = [
        r'\bjest\b',
        r'\bmocha\b',
        r'\btap\b',
        r'\bpytest\b',
        r'\bnpm\s+(run\s+)?test\b',
        r'\bnode\s+--test\b',
        r'\bpython\s+-m\s+pytest\b',
    ]

    for pattern in real_test_patterns:
        if re.search(pattern, script):
            return True

    # If we get here, it's unclear - assume it's a real test
    # (could be a custom script like './run-tests.sh')
    return True
